Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp

```diff
@@ -276,6 +276,42 @@ void RemoteGraphicsContextGL::simulateEventForTesting(WebCore::GraphicsContextGL
     m_context->simulateEventForTesting(event);
 }
 
+void RemoteGraphicsContextGL::getBufferSubDataInline(uint32_t target, uint64_t offset, size_t dataSize, CompletionHandler<void(std::span<const uint8_t>)>&& completionHandler)
+{
+    assertIsCurrent(workQueue());
+    static constexpr size_t getBufferSubDataInlineSizeLimit = 64 * KB; // NOTE: when changing, change the value in RemoteGraphicsContextGLProxy too.
+    if (!dataSize) {
+        completionHandler({ });
+        return;
+    }
+    MallocPtr<uint8_t> bufferStore;
+    std::span<uint8_t> bufferData;
+    if (dataSize < getBufferSubDataInlineSizeLimit) {
+        bufferStore = MallocPtr<uint8_t>::tryMalloc(dataSize);
+        if (bufferStore)
+            bufferData = { bufferStore.get(), dataSize };
+    }
+    std::span<const uint8_t> validBufferData;
+    if (bufferData.size() == dataSize)
+        validBufferData = m_context->getBufferSubDataWithStatus(target, offset, bufferData);
+
+    completionHandler(bufferData);
+}
+
+void RemoteGraphicsContextGL::getBufferSubDataSharedMemory(uint32_t target, uint64_t offset, WebCore::SharedMemory::Handle handle, CompletionHandler<void(size_t)>&& completionHandler)
+{
+    assertIsCurrent(workQueue());
+    std::span<const uint8_t> validBufferData;
+
+    handle.setOwnershipOfMemory(m_sharedResourceCache->resourceOwner(), WebKit::MemoryLedger::Default);
+    if (auto buffer = SharedMemory::map(WTFMove(handle), SharedMemory::Protection::ReadWrite))
+        validBufferData = m_context->getBufferSubDataWithStatus(target, offset, buffer->mutableSpan());
+    else
+        m_context->addError(GCGLErrorCode::InvalidOperation);
+
+    completionHandler(validBufferData.size());
```

Comment:
the two cases: inline and shm cannot produce different results o errors.
Currently:
- inline: the memory contains zeros for errors
- shm: returns the buffer size that is valid

So these should somehow be behaving the same

Note: It may be that previous getBufferSubData was copying over the client buffer even when the GPUP would run into an error.